### DISCUSSION TITLE
how to rename JSON object keys

### DISCUSSION THREAD

**Discussion by nsitbon** [ohler55/ojg#172] [Q&A] [answered]
Hi after looking at the documentation I can't find a way to rename JSON object keys. Can you show me how to do that please.

Best regards

**Comment by ohler55** [comment]
In go there is no way to rename a map key. You can, however, add and delete keys so deleting a key with `Del` and then `SetOne` will have the same effect. Be aware that if calling either during an iteration of the map may rearrange the map leaving some keys untouched.

**Comment by nsitbon** [comment]
Thank you for your answer : I'm not talking about Go here but about the possibilities offered by your library. I'm receiving a deep JSON object inside one of my endpoint and I want to rename some specific keys. It's very easy to change object values using your library but I'm struggling to change object keys. 
Best regards

**Reply by ohler55** [reply]
Before any changes can be made the JSON has to be parsed into data structures. Since OjG is written in go it parses to go data structures. In the general cal JSON objects become `map[string]any` in go. That is what the jp package in OjG has to work with. Changes can be made and the data written as JSON but OjG does not operate on or modify the raw text which brings us back to attempting to modify go data with the limitations imposed by the go language. 

Maybe a small example of what you are trying to achieve would help.

**Reply by nsitbon** [reply]
I'm trying to rename an arbitrary key inside a JSON value.
I have a bunch of endpoints all returning different JSON value (can be an array of the JSON object).
Our users will provide a JSONPATH matching sub objects, an old key name and a new key name.
So for example given the following JSON value:
```json
{
    "firstName": "John",
    "lastName": "doe",
    "age": 26,
    "address": {
        "streetAddress": "naist street",
        "city": "Nara",
        "postalCode": "630-0192",
        "other": {
            "type": 456
        }
    },
    "phoneNumbers": [
        {
            "type": "iPhone",
            "number": "0123-4567-8888"
        },
        {
            "type": "home",
            "number": "0123-4567-8910"
        }
    ]
}
```
and the following JSONPATH : `$..*[?(@.type)]`
old key : type 
new key : newType
we want to rewrite it like that:
```json
{
    "firstName": "John",
    "lastName": "doe",
    "age": 26,
    "address": {
        "streetAddress": "naist street",
        "city": "Nara",
        "postalCode": "630-0192",
        "other": {
            "newType": 456
        }
    },
    "phoneNumbers": [
        {
            "newType": "iPhone",
            "number": "0123-4567-8888"
        },
        {
            "newType": "home",
            "number": "0123-4567-8910"
        }
    ]
}
```

**Reply by thadguidry** [reply]
@nsitbon You might take a look at Polars which can run with Python or Rust.  It can [read/write JSON](https://docs.pola.rs/user-guide/io/json/) and you can use pattern matching to find the keys that you wish to replace.
https://docs.pola.rs/user-guide/expressions/strings/#replace-a-pattern
For further help, you can ask their Discord community which there are links at the top of their [README](https://github.com/pola-rs/polars) in their GitHub.

There's a few other tools (ask around on Stackoverflow - or search across GitHub - I'm [not "quite" sure](https://duckdb.org/docs/extensions/json.html#transforming-json) if you could pull this off with DuckDB even. dunno.) that could do the key replacement within JSON, but I think you might like Polars particularly for this need.

@ohler55 is there truly no way for something like what Polars has to be implemented in OjG (given enough time/money of course) ?

**Reply by thadguidry** [reply]
@nsitbon Or you could [flatten "really really fast"](https://github.com/vtselfa/json-objects-to-csv) (perhaps a smidgen faster than OjG flatten - dunno) and rename within CSV field processing with your favorite tools?

**Reply by nsitbon** [reply]
Hi @thadguidry thank you for all your propositions but ideally I would like to stick to one lib in Golang...

**Reply by thadguidry** [reply]
1 lib to rule all your JSON possible transforms?  I don't think even Peter has set THAT expectation (yet) with OjG.  Good luck sir.  **I'll be interested in his response to this.**  I think it could be done.  Perhaps instead of extracting values one-by-one, instead for the "replace" function it could extract all values at once and replace them. Dunno.

**Reply by ohler55** [reply]
With a non deterministic path like the one given the best approach is probably to use `jp.Locate` to get a set of normalized paths and then for each one get the value, delete the element at the path, and change the path to be the new version.

**Reply by ohler55** [reply]
Would code for the solution described be useful?

**Reply by nsitbon** [reply]
Hi @ohler55 I was able to write a working version using your hints but I'm far from what I thought I could do. This is the first time I'm dealing with JSONPATH and your library and even JSONPATH is very limited. The first 2 features I was looking for were 
- be able to rename a JSON object key
- be able to apply a transformation on any JSON object matching arbitrary key inside a JSON object

it turns out that both aren't handled natively (for the second see https://stackoverflow.com/a/78445097/24913827).
I still would like to thank you for your precious time and help. I'm still hesitating between trying to make things work with JSONPATH and your library OR trying to make a new library with a new query syntax... 

Here is the code for those who would be interested
```golang
func renameKey(jsValue any, oldKey, newKey string) error {
	x, err := jp.ParseString("$.." + oldKey)

	if err != nil {
		return err
	}

	exprs := x.Locate(jsValue, 0)

	matchingObjects, err := slice.MapWithErr(exprs, func(expr jp.Expr) (map[string]any, error) {
		if y, err := jp.ParseString(strings.TrimSuffix(expr.String(), "."+oldKey)); err != nil {
			return nil, err
		} else {
			return y.Get(jsValue)[0].(map[string]any), nil
		}
	})

	if err != nil {
		return err
	}

	slice.Foreach(matchingObjects, func(m map[string]any) {
		old := m[oldKey]
		delete(m, oldKey)
		m[newKey] = old
	})

	return nil
}

func TestRenameKey(t *testing.T) {
	data, err := oj.ParseString(`{"a":{"a": 1, "b": {"a": 2}}, "b": 4}`)
	require.NoError(t, err)

	err = renameKey(data, "a", "c")
	require.NoError(t, err)

	r, _ := json.Marshal(data)
	require.JSONEq(t, `{"c":{"c": 1, "b": {"c": 2}}, "b": 4}`, string(r))
}

```

**Reply by ohler55** [reply]
It's a bit round about but if it works for you that's good. You could simplify and improve performance it by using this function:

```golang
func RenameKeys(data any, oldKey, newKey string) error {
	x := jp.D().C(oldKey)
	for _, np := range x.Locate(data, 0) {
		value := np.First(data)
		_ = np.Del(data)
		np[len(np)-1] = jp.Child(newKey)
		if err := np.SetOne(data, value); err != nil {
			return err
		}
	}
	return nil
}
```

**Reply by nsitbon** [reply]
I must admit I'm not confortable with this line `np[len(np)-1] = jp.Child(newKey)` : np seems to encapsulate the behaviour of a queue where you're calling the First method and then you go back to standard Golang subscripting operator to mutate the last element... doesn't really makes sense to me but I have to trust you ;-)

**Reply by ohler55** [reply]
If you look at the jp.Expr you will see it is a slice of jp.Frags. It perfectly reasonable to replace elements of the slice. Converting the slice to a string, replacing the string suffix, and then converting back to a slice is more error prone and very inefficient.

**Comment by ohler55** [comment, marked as answer]
```golang
func TestRename(t *testing.T) {
	data := oj.MustParse([]byte(`
{
    "firstName": "John",
    "lastName": "doe",
    "age": 26,
    "address": {
        "streetAddress": "naist street",
        "city": "Nara",
        "postalCode": "630-0192",
        "other": {
            "type": 456
        }
    },
    "phoneNumbers": [
        {
            "type": "iPhone",
            "number": "0123-4567-8888"
        },
        {
            "type": "home",
            "number": "0123-4567-8910"
        }
    ]
}
`))
	path := jp.MustParseString("$..type")
	fmt.Printf("before: %s\n", pretty.SEN(data))
	for _, np := range path.Locate(data, -1) {
		value := np.First(data)
		np.MustDel(data)
		np[len(np)-1] = jp.Child("newType")
		np.MustSetOne(data, value)
	}
	fmt.Printf("after: %s\n", pretty.SEN(data))
}
```
